### PR TITLE
Add "Exact mechanism" section to the derivation page

### PR DESCRIPTION
**Pull request by Sporarum**
Specifies the steps taken by automatic derivation

Might get removed again once they are incorporated in the spec

### CODE REVIEW

File: docs/_docs/reference/contextual/derivation.md
Comment (julienrf) [inline docs/_docs/reference/contextual/derivation.md:97]: Here too, at least one example would be welcome.

File: docs/_docs/reference/contextual/derivation.md
Comment (julienrf) [inline docs/_docs/reference/contextual/derivation.md:120]: Could you show what the error looks like?

File: docs/_docs/reference/contextual/derivation.md
Comment (julienrf) [inline docs/_docs/reference/contextual/derivation.md:102]: So, `derives TC` works only if `TC` takes 1 parameter? (except `CanEqual`) It would be good to highlight this information better, I think, in the introductory paragraph (e.g. by saying “A type class in this sense is any trait or class with a _single_ type parameter”).

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [inline docs/_docs/reference/contextual/derivation.md:44]: debatable but its probably better to be explicit then imply hidden context bounds
```suggestion
given [...](using ...): TC[ ... DerivingType[...] ... ] = TC.derived
```

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md:52]: It felt weird to put it in the title, so I clarified the body

Let me know if it's okay like that

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md]: I added an example error, do you think it's okay like that ?

File: docs/_docs/reference/contextual/derivation.md
Comment (julienrf) [reply docs/_docs/reference/contextual/derivation.md:52]: Thank you!

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [inline docs/_docs/reference/contextual/derivation.md:44]:
```suggestion
given [...]: TC[[...] =>> DerivingType[...]] = TC.derived
```

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md]: I changed the intro, and also renamed this section, hopefully it should be clearer now

File: docs/_docs/reference/contextual/derivation.md
Comment (julienrf) [reply docs/_docs/reference/contextual/derivation.md]: Yes, even though it’s unfortunate that we can’t be more specific about the error message.

File: docs/_docs/reference/contextual/derivation.md
Comment (julienrf) [reply docs/_docs/reference/contextual/derivation.md]: Great!

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md:44]: I see what you mean, but I think it would be more confusing than anything

Maybe I should add a note in the part where the context bounds are used ?

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md]: I applied it, just not as a note, as I feel like this adds new information

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md:111]: I couldn't find a realistic, self-contained example
Given how advanced this part of derivation is, my intuition is that it's okay to not give an example ?

My goal would be to have the majority of the "mass" of this section be on the common cases: `CanEqual` and the one with context bounds

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [inline docs/_docs/reference/contextual/derivation.md:94]:
```suggestion
If `F` and `DerivingType` take the same number of arguments (`N == K`):
```

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [inline docs/_docs/reference/contextual/derivation.md:108]:
```suggestion
If `F` takes less arguments than `DerivingType` (`K < N`), we fill in the leftmost slots with type parameters:
```

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md:58]: The problem is that "`F` and all arguments of `DerivingType` have kind `*`" is not equivalent to "`K == 0`", but "i.e." would imply that

I can't find a way of saying it that is neither clunky nor containing that implication

File: docs/_docs/reference/contextual/derivation.md
Comment (julienrf) [reply docs/_docs/reference/contextual/derivation.md:111]: OK

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [reply docs/_docs/reference/contextual/derivation.md]: not sure if this wording is perfect

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [reply docs/_docs/reference/contextual/derivation.md:58]: well I didn't mean it to be "equivilancy", just that the title implies K == 0

Comment (bishabosha) [reply docs/_docs/reference/contextual/derivation.md:58]:
```suggestion
**Note:** `K == 0` in this case
```

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [inline docs/_docs/reference/contextual/derivation.md:108]:
```suggestion
If `F` takes less arguments than `DerivingType` (`K < N`), we fill in the remaining leftmost slots with type parameters of the given:
```

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [inline docs/_docs/reference/contextual/derivation.md:137]:
```suggestion
  given [A_L, A_R, G_L[_], G_R[_]](using CanEqual[A_L, A_R]): CanEqual[MyClass[A_L, G_L], MyClass[A_R, G_R]] = CanEqual.derived
```

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [inline docs/_docs/reference/contextual/derivation.md:117]:
```suggestion
Let `U_1`, ..., `U_M` be the parameters of `DerivingType` of kind `*`. (`U_i ∈ {T_i}`) 
```

**Comment by bishabosha** [comment]
I'm ok except for my comments

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md:117]: What about:
```suggestion
Let `U_1`, ..., `U_M` be the parameters of `DerivingType` of kind `*`. (These are a subset of the `T_i`s)
```

**Comment by Sporarum** [comment]
Thank you for your reviews !

I will merge this PR myself after having rebased the commits

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md:44]: What about the following ?
Adding the using clause to the "general shape", and 
Adding a note to the section with the context bounds reminding they do indeed introduce a using

File: docs/_docs/reference/contextual/derivation.md
Comment (sjrd) [reply docs/_docs/reference/contextual/derivation.md:59]: `i.e.` stands of "id est" in Latin, which means "that is". So yes, it means equivalence.

`e.g.` stands for "exempli gratia", which literally means "for example".

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [reply docs/_docs/reference/contextual/derivation.md:118]: looks good!

File: docs/_docs/reference/contextual/derivation.md
Comment (bishabosha) [reply docs/_docs/reference/contextual/derivation.md:44]: or you can reference the expansion of context bounds from here https://docs.scala-lang.org/scala3/reference/contextual/context-bounds.html

**Review by bishabosha — APPROVED** [review]

File: docs/_docs/reference/contextual/derivation.md
Comment (Sporarum) [reply docs/_docs/reference/contextual/derivation.md]: I just did this, literally this second, what a coincidence

I'll let you look and tell me what you think

Comment (bishabosha) [reply docs/_docs/reference/contextual/derivation.md]: yeah it looks good!

**Pull request merged**
